rfcs/20200809-keras-saved-model-update.md

```diff
@@ -0,0 +1,175 @@
+
+When loading with `LoadOptions(nodes_to_load=["root.child"])`, the loader will return a dict `{"root.child": UserObject}`. The object will have the `a`, `b`, and `call` attributes.
+
+Alternatively, say the user has already created the `child` object, with the `a` variable but no `b` variable:
+```
+child_obj = tf.Module()
+a_var = child_obj.a = tf.Variable(5.)
+tf.saved_model.load(
+    path, 
+    options=tf.saved_model.LoadOptions(
+        nodes_to_load={"root.child": child_obj}))
+print(a_var is child_obj.a)  # True
+print(child_obj.b)  # tf.Variable
+```
+The loader will create the b variable but will not overwrite the a variable.
+
+### Updated Keras Serialization
+To summarize, Keras SavedModels will be produced and loaded following these steps:
+
+*Saving*
+
+1. Call the core SavedModel saver to save the TensorFlow object graph and tf.functions.
+2. Write a `keras.metadata` file to the directory, containing data for all Keras object nodes.
+
+*Loading*
+
+1. Read the `keras.metadata` file (or metadata from the SavedModel proto if the file is not present).
+2. Create all Keras objects. If the config is available for an object, call the Keras layer deserializer. If the config is not available, then create a RevivedLayer object.
+3. Call the core SavedModel loader using the `nodes_to_load` option. The loader will not overwrite the already-created objects and variables.
+
+#### Benchmarks
+With the modified implementation, saving takes a bit more time in order to generate the Keras metadata file, while loading is a lot faster. These measurements are taken from the Keras applications SavedModel benchmarks.
+
+![saving](20200809-keras-saved-model-update/saving.png)
+![loading](20200809-keras-saved-model-update/loading.png)
+
+### Alternatives Considered
+An alternative that was long considered was having the user register serializer and deserializer functions to each class that should be specially handled during SavedModel serialization and deserialization. However, this idea was put to rest due to the recursive nature of the TensorFlow object graph and the complicated issues [mentioned above](#why-separate-core-vs-high-level-serialization).
+
+### Breaking changes
+`tf.saved_model.save` will no longer save the metadata added to Keras models. With this, only the Keras `model.save` function will create SavedModels that are compatible with the Keras model loader `tf.keras.models.load_model`. When using the core loader (`tf.saved_model.load`), an object with specific endpoints is loaded. Please see the Keras SavedModel RFC for the list of all attributes that are stored: https://github.com/k-w-w/community/blob/master/rfcs/20190509-keras-saved-model.md#serialization-details
+
+| Saver  | Loader |  |
+| ---- | ---- | ---- | 
+| `tf.saved_model.save` | `tf.saved_model.load` | No change |
+| `tf.keras.models.save_model` | `tf.saved_model.load` | No change |
+| `tf.saved_model.save` | `tf.keras.models.load_model` | **Error** |
```

Comment:
Does it mean tf.saved_model.save() will no longer trace each layer call() and enforce requirements?
Is there a support for multiple input arguments to model and layer?

Reply:
tf.saved_model will still save each layer call by default -- only the metadata is removed. The bug involving multiple input arguments is being looked at but not in the scope of this doc